protocols/gossipsub/src/behaviour/tests.rs

```diff
@@ -1852,6 +1852,65 @@ fn test_connect_to_px_peers_on_handle_prune() {
     ));
 }
 
+#[test]
+fn test_connect_to_px_peers_with_peer_record_on_handle_prune() {
+    let config: Config = Config::default();
+
+    let (mut gs, peers, topics) = inject_nodes1()
+        .peer_no(1)
+        .topics(vec!["test".into()])
+        .to_subscribe(true)
+        .create_network();
+
+    // handle prune from single peer with px peers
+
+    let mut px_peer_ids = Vec::new();
+    let mut px = Vec::new();
+    // propose more px peers than config.prune_peers()
+    for i in 0..config.prune_peers() + 5 {
+        let key = Keypair::generate_ed25519();
+        let address: Multiaddr = format!("/ip4/1.2.3.4/tcp/{i}").try_into().unwrap();
+        let peer_record = PeerRecord::new(&key, vec![address]).unwrap();
+
+        px_peer_ids.push(key.public().to_peer_id());
+        px.push(PeerInfo {
+            peer_id: None,
+            signed_peer_record: Some(peer_record.into_signed_envelope()),
+        });
+    }
+
+    gs.handle_prune(
+        &peers[0],
+        vec![(
+            topics[0].clone(),
+            px.clone(),
+            Some(config.prune_backoff().as_secs()),
+        )],
+    );
+
+    // Check DialPeer events for px peers
+    let dials: Vec<_> = gs
+        .events
+        .iter()
+        .filter_map(|e| match e {
+            // TODO: How to extract addresses from DialOpts to assert them in the test?
+            NetworkBehaviourAction::Dial { opts } => opts.get_peer_id(),
+            _ => None,
+        })
+        .collect();
+
+    // Exactly config.prune_peers() many random peers should be dialled
+    assert_eq!(dials.len(), config.prune_peers());
+
+    let dials_set: HashSet<_> = dials.into_iter().collect();
+
+    // No duplicates
+    assert_eq!(dials_set.len(), config.prune_peers());
+
+    // all dial peers must be in px
+    assert!(dials_set.is_subset(&px_peer_ids.iter().copied().collect::<HashSet<_>>()));
```

Comment:
I think it would be good to assert the presence of the addresses too!

Reply:
Any suggestion on how to retrieve the addresses from DialOpts struct? 

https://github.com/libp2p/rust-libp2p/blob/3f5dd631aef6592fbcb1439bdd311ae9f81be684/protocols/gossipsub/src/behaviour/tests.rs#L1896-L1897

Or should I make this fn public?

https://github.com/libp2p/rust-libp2p/blob/d81f9476f72a6eb1d4bc683ad04b3fdf866cfc12/swarm/src/dial_opts.rs#L130-L132

Reply:
I am not really a fan of the style of the tests in gossipsub anyway. We should only be testing behaviours by running them in a `Swarm` and not call these lifecycle functions manually.

We now have a new crate, `libp2p-swarm-test` which makes it easy to set up swarms for testing.

How about we do the following:

- Implement "sending" of signed peer records by keeping a hashmap of records in the behaviour where we perform a lookup upon prune
- At the moment, we don't have a way of adding records to this hashmap so in reality, there won't ever be any
- Within the test, we can add a record (by accessing the field directly)
- We can then write a test using `libp2p-swarm-test` that ends up sending a `Prune` to a node which will then contain the `SignedPeerRecord`
- We can then assert that the `Swarm` establishes a connection to the node specified in the prune message

What do you think?

Reply:
I am not really a fan of the style of the tests in gossipsub anyway. We should only be testing behaviours by running them in a `Swarm` and not call these lifecycle functions manually.

We now have a new crate, `libp2p-swarm-test` which makes it easy to set up swarms for testing.

How about we do the following:

- Implement "sending" of signed peer records by keeping a hashmap of records in the behaviour where we perform a lookup upon prune
- At the moment, we don't have a way of adding records to this hashmap so in reality, there won't ever be any
- Within the test, we can add a record (by accessing the field directly)
- We can then write a test using `libp2p-swarm-test` that ends up sending a `Prune` to a node which will then contain the `SignedPeerRecord`
- We can then assert that the `Swarm` establishes a connection to the node specified in the prune message

What do you think?

Reply:
going this way seems reasonable to me

Reply:
That makes tho if all tests in this package are written with a consistent style, should this PR stick to that style? And then refactor them into using swarm-test in another PR

Reply:
> That makes tho if all tests in this package are written with a consistent style, should this PR stick to that style? And then refactor them into using swarm-test in another PR

Eventually yes but that is a massive effort. I don't mind mixing styles in a transition period if we get a better, more feature-oriented test in exchange.